foundry.md

```diff
@@ -373,18 +373,67 @@ This rule takes the `address` value using `#asWord(#range(LM, ARGSTART +Int 4, 3
-          <output> _ => #bufStrict(32, NONCE) </output>
+          <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH RETSTART RETWIDTH
+           => #loadAccount #asWord(#range(LM, ARGSTART +Int 4, 32))
+           ~> #foundryGetNonce ARGSTART RETSTART RETWIDTH
+           ~> #refund GCALL ...
+          </k>
+          <callGas> GCALL => 0 </callGas>
           <localMem> LM </localMem>
-          <account>
+       requires CHEAT_ADDR ==Int #address(FoundryCheat)
+        andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 755185067 // selector ( "getNonce(address)" )
+     [priority(40)]
+
+
+     syntax KItem ::= "#foundryGetNonce" Int Int Int
+ // ----------------------------------------------------------
+    rule <k> #foundryGetNonce ARGSTART RETSTART RETWIDTH
+          => #setLocalMem RETSTART RETWIDTH #bufStrict(32, NONCE)
+          ~> 1 ~> #push ...
+         </k>
+         <output> _ => #bufStrict(32, NONCE) </output>
+         <account>
              <acctID> ACCTID </acctID>
-             <nonce>  NONCE  </nonce>
+             <nonce>  NONCE </nonce>
              ...
          </account>
+         <localMem> LM </localMem>
+      requires ACCTID ==Int #asWord(#range(LM, ARGSTART +Int 4, 32))
+```
+
+#### `setNonce` - Sets the nonce of the given account.
+
+```
+function setNonce(address account, uint64 nonce) external;
+```
+
+`call.setNonce` will match when the `setNonce` function is called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
+This rule takes the `address` value using `#asWord(#range(LM, ARGSTART +Int 4, 32)` and `uint64` value corresponding to new nonce using `#asWord(#range(LM, ARGSTART +Int 36, 32)` and updates the `<nonce>` cell in the respective account.
+
+```k
+    rule [call.setNonce]:
+          <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH _RETSTART _RETWIDTH
+           => #loadAccount #asWord(#range(LM, ARGSTART +Int 4, 32))
+           ~> #foundrySetNonce ARGSTART
+           ~> #refund GCALL ...
+          </k>
+          <localMem> LM </localMem>
+          <callGas> GCALL => 0 </callGas>
        requires CHEAT_ADDR ==Int #address(FoundryCheat)
-        andBool ACCTID ==Int #asWord(#range(LM, ARGSTART +Int 4, 32))
-        andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 755185067 // selector ( "getNonce(address)" )
+        andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 4175530839 // selector ( "setNonce(address,uint64)" )
      [priority(40)]
+
+     syntax KItem ::= "#foundrySetNonce" Int
+ // ----------------------------------------------------------
+    rule <k> #foundrySetNonce ARGSTART => 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <account>
+             <acctID> ACCTID </acctID>
+             <nonce>  _ => #asWord(#range(LM, ARGSTART +Int 36, 32))  </nonce>
+             ...
+         </account>
+         <localMem> LM </localMem>
+      requires ACCTID ==Int #asWord(#range(LM, ARGSTART +Int 4, 32))
 ```
```

Comment:
Same here, pre-compute the ACCTID in th efirst rule, then use it directly in the second rule. This is the difference between `O(N)` in the size of the accounts cell, and `O(log(N))`.

Reply:
Thanks! I didn't know (or remember) that. It's fixed now. Please take a look.